framework/src/modules/chain/blocks/fork_choice_rule.js

```diff
@@ -1,3 +1,59 @@
+/**
+ * Utility functions only required in this file.
+ */
+
+/**
+ * Wrapper for readability. Returns the forging slot of a block
+ * @param slots
+ * @param block
+ * @return {number}
+ */
+const forgingSlot = (slots, block) => slots.getSlotNumber(block.timestamp);
+
+/**
+ * Returns whether the last block applied (tip of the chain) was received withing its
+ * designated forging slot.
+ *
+ * IMPORTANT: If the last block applied on the chain was not received from the network
+ * but instead it was explicitly synced or forged, the return value of this function is true
+ * This is evaluated in the first line.
+ * @param slots
+ * @param lastAppliedBlock
+ * @param lastReceivedAndAppliedBlock
+ * @return {boolean}
+ */
+const islastAppliedBlockReceivedWithinForgingSlot = (
+	slots,
+	lastAppliedBlock,
+	lastReceivedAndAppliedBlock
+) => {
+	if (lastAppliedBlock.id !== lastReceivedAndAppliedBlock.id) return true;
+
+	return isBlockReceivedWithinForgingSlot(
+		slots,
+		lastAppliedBlock,
+		lastReceivedAndAppliedBlock.timestamp
+	);
+};
+
+/**
+ * Wrapper for readability. Returns whether a block was received in its designated
+ * forging slot.
+ * @param slots
+ * @param receivedBlock
+ * @param receivedTime
+ * @return {boolean}
+ */
+const isBlockReceivedWithinForgingSlot = (slots, receivedBlock, receivedTime) =>
```

Comment:
If full block not required, suggest to only pass the required attribute as `blockTime`